libs/core/src/main/java/org/opensearch/core/common/io/stream/BaseWriteable.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.core.common.io.stream;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Implementers can be written to a {@code StreamOutput} and read from a {@code StreamInput}. This allows them to be "thrown
+ * across the wire" using OpenSearch's internal protocol. If the implementer also implements equals and hashCode then a copy made by
+ * serializing and deserializing must be equal and have the same hashCode. It isn't required that such a copy be entirely unchanged.
+ *
+ * @opensearch.internal
+ */
+public interface BaseWriteable<S extends BaseStreamOutput> {
+    /**
+     * A WriteableRegistry registers {@link Writer} methods for writing data types over a
+     * {@link BaseStreamOutput} channel and {@link Reader} methods for reading data from a
+     * {@link BaseStreamInput} channel.
+     *
+     * @opensearch.internal
+     */
+    class WriteableRegistry {
```

Comment:
I suspect we the registry could be accessed from multiple threads, `ConcurrentHashMap` (or synchronized map) is safer option here

Reply:
Good call.. switched to concurrent.